File: frontend-src/src/services/palette.ts
import type { PaletteId } from '../types'

export interface PaletteColor {
  value: string
  label: string
}

const BASIC_COLORS = {
  black: { value: '#000000', label: 'Black' },
  white: { value: '#ffffff', label: 'White' },
  red: { value: '#d22626', label: 'Red' },
  yellow: { value: '#e5b800', label: 'Yellow' },
  blue: { value: '#285995', label: 'Blue' },
  green: { value: '#72a85a', label: 'Green' },
} as const

const grayscale = (levels: number): PaletteColor[] => Array.from({ length: levels }, (_, index) => {
  const channel = Math.round((255 * index) / (levels - 1))
  const value = `#${channel.toString(16).padStart(2, '0').repeat(3)}`
  return { value, label: index === 0 ? 'Black' : index === levels - 1 ? 'White' : `Gray ${index}` }
})

export const PALETTE_COLORS: Record<PaletteId, PaletteColor[]> = {
  bw: [BASIC_COLORS.black, BASIC_COLORS.white],
  gray4: grayscale(4),
  gray16: grayscale(16),
  bwr: [BASIC_COLORS.black, BASIC_COLORS.white, BASIC_COLORS.red],
  bwy: [BASIC_COLORS.black, BASIC_COLORS.white, BASIC_COLORS.yellow],
  bwry: [BASIC_COLORS.black, BASIC_COLORS.white, BASIC_COLORS.red, BASIC_COLORS.yellow],
  spectra6: [
    BASIC_COLORS.black,
    BASIC_COLORS.white,
    BASIC_COLORS.red,
    BASIC_COLORS.yellow,
    BASIC_COLORS.blue,
    BASIC_COLORS.green,
  ],
}

export const paletteColors = (palette: PaletteId): PaletteColor[] => PALETTE_COLORS[palette]

export const resolvePaletteColor = (palette: PaletteId, value: unknown): string => {
  const colors = paletteColors(palette)
  const normalized = String(value ?? '').toLowerCase()
  return colors.find((color) => color.value === normalized)?.value ?? colors[0].value
}

File: custom_components/opendisplay_studio/palette.py
"""Canonical output colors exposed to palette-aware widget fields."""

from __future__ import annotations

from typing import Final


def _grayscale(levels: int) -> tuple[str, ...]:
    """Return evenly distributed grayscale colors including black and white."""
    values = (round(255 * index / (levels - 1)) for index in range(levels))
    return tuple(f"#{channel:02x}{channel:02x}{channel:02x}" for channel in values)


PALETTE_COLORS: Final[dict[str, tuple[str, ...]]] = {
    "bw": ("#000000", "#ffffff"),
    "gray4": _grayscale(4),
    "gray16": _grayscale(16),
    "bwr": ("#000000", "#ffffff", "#d22626"),
    "bwy": ("#000000", "#ffffff", "#e5b800"),
    "bwry": ("#000000", "#ffffff", "#d22626", "#e5b800"),
    "spectra6": (
        "#000000",
        "#ffffff",
        "#d22626",
        "#e5b800",
        "#285995",
        "#72a85a",
    ),
}


def normalize_palette_color(palette: str, value: object) -> str:
    """Keep a configured color only when the selected display can produce it."""
    colors = PALETTE_COLORS[palette]
    normalized = str(value).lower()
    return normalized if normalized in colors else colors[0]
